core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -42,29 +42,45 @@ namespace kagome::blockchain {
 
   boost::optional<std::vector<std::shared_ptr<BlockTreeImpl::TreeNode>>>
   BlockTreeImpl::TreeNode::getPathTo(const primitives::BlockHash &hash) {
-    std::vector<std::shared_ptr<TreeNode>> stack;
-    std::vector<std::shared_ptr<TreeNode>> to_check;
-    to_check.emplace_back(shared_from_this());
+    std::vector<std::shared_ptr<TreeNode>> stack{shared_from_this()};
+    std::unordered_map<primitives::BlockNumber, size_t> forks;
 
-    do {
-      auto target = to_check.back();
-      to_check.pop_back();
+    bool came_back = false;
+    while (not stack.empty()) {
+      auto &current = stack.back();
 
-      if (target->block_hash == hash) {
-        stack.emplace_back(target);
+      // Found
+      if (current->block_hash == hash) {
         return stack;
       }
 
-      auto parent = target->parent.lock();
-      while (!stack.empty() && parent != stack.back()) {
-        stack.pop_back();
+      // Straight
```

Comment:
You need just change places: first clear loop, than if-block. You don't need hash-table here at all:
```c++
      auto parent = target->parent.lock();
      while (!stack.empty() && parent != stack.back()) {
        stack.pop_back();
      }

      if (target->block_hash == hash) {
        stack.emplace_back(target);
        return stack;
      }